### ISSUE TITLE
Exception control: “try” should rethrow flow control exceptions

### ISSUE THREAD

**Issue by hanche** [elves/elvish#1396]
In the following toy example, I had naively expected the `return` command to result in a return from `foo`. But instead, `try` catches the exception, as it is designed to do. But I think it is a bit too eager.

```
⬥ fn foo { try { put a; return } except e { put $e }; put b }
⬥ foo
⮕ a
⮕ [&reason=[&name=return &type=flow]]
⮕ b
```

Suggestion: After executing the `finally` block, if any, the exception should be rethrown if it was of a flow-control type.

**Comment by krader1961** [comment]
This will clearly need some discussion and careful thought before, or even if, the behavior is changed. My feeling is that the builtin flow control exceptions raised by `return`, `break` and `continue` should not be caught by `try`. The exceptions raised by those commands are fundamentally different in nature from an exception that occurs, for example, due to an external command exiting with a non-zero status or passing invalid arguments to a builtin.

**Comment by hanche** [comment]
One question that will need answering: Should any `else` or `finally` block be executed it the `try` block throws a control flow exception? An argument in favour is that these blocks are most often used to clean up after actions that may have been initiated in the `try` block. But then the exception has to be caught and rethrown. The counterargument is, I think, that it is up to the user to ensure proper cleaning up if they escape via flow control. And the countercounterargument: The user might well expect these blocks will take care of that. I see no totally decisive argument either way.

**Comment by krader1961** [comment]
> Should any `else` or `finally` block be executed it the try block throws a control flow exception?

Yes. If I'm executing a pipeline in a `try` block where I want to perform a `break` or `continue` if there are no exceptions I **do not** want to have to test for control flow exceptions in the `except` block and that block should not be run for those exceptions.

On the other hand an `else` block is only executed today if no exception occurred. I think we **do** want to change that behavior to run the `else` block when a control flow exception occurs; i.e., behave as if no exception occurred and automatically re-raise the control flow exception.

The `finally` block case would be unchanged. Whether or not control flow exceptions are special-cased  it would still be run when an uncaught control flow exception occurs.
